libs/utils/include/utils/JobStorage.h

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_UTILS_JOBSTORAGE_H
+#define TNT_UTILS_JOBSTORAGE_H
+
+#include <utils/FixedCapacityVector.h>
+#include <utils/Invocable.h>
+#include <utils/JobSystem.h>
+
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+#include <memory>
+#include <unordered_map>
+#include <limits>
+#include <queue>
+
+namespace utils {
+
+/**
+ * A thread-safe producer-consumer queue with batching capabilities.
+ *
+ * This class is thread-safe. All public methods can be called from any thread.
+ *
+ * This class is stateless regarding concurrency. The *caller* decides the blocking behavior and/or
+ * batching when they call a 'pop' methods.
+ *
+ * A typical use case looks like this:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * #include <utils/JobSystem.h>
+ * using namespace utils;
+ *
+ * JobStorage::Ptr storage = JobStorage::create();
+ * JobWorker::Ptr worker = AmortizationWorker::create(storage);
+ * [ or JobWorker::Ptr worker = ThreadWorker::create(storage, config); ]
+ *
+ * void loop() {
+ *     worker->process(2); // for AmortizationWorker
+ * }
+ *
+ * void cleanup() {
+ *     worker->terminate();
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * JobId id = storage->push([](){ ... });
+ * storage->cancel(id);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * JobId preIssuedId = storage->issueJobId();
+ * JobId id = storage->push([](){ ... }, preIssuedId);
+ * assert(id == preIssuedId);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ */
+class JobStorage : public std::enable_shared_from_this<JobStorage> {
```

Comment:
We already have something called `JobSystem`, but this is different and more specialized. So a few things come to mind:

1. should this be in libutils? (meaning it's essentially a public API). Could it be just an implementation detail in libbackend?

2. if it's in libutils, we need to be more careful about naming. It looks like it could be related to the `JobSystem` but it's not. And `JobStorage` is a bit odd, maybe it should be `JobQueue`.

Personally I think it shouldn't be in libutils, even if it's fairly generic, it's using a lot of STL headers we usually don't use un public headers.